### ISSUE TITLE
This should become part of DrRacket

### ISSUE THREAD

**Issue by spdegabrielle** [MatrixForChange/files-viewer#32]
I think this plugin should become part of DrRacket included in the main distribution.

Is this something you would be interested in?

A requirement would be for this project to use a compatible licence. (MIT/Apache)

See contributing at https://github.com/racket/drracket/blob/master/README.md

Please let me know what you think?

Kind regards

Stephen

**Comment by Syntacticlosure** [comment]
I have no problem with relicensing. What about @yjqww6 , @vityou , and @Aloxaf ?

**Comment by yjqww6** [comment]
ok

**Comment by Aloxaf** [comment]
ok

**Comment by zlee-personal** [comment]
Sure

**Comment by spdegabrielle** [comment]
PR created - 'change license to MIT & Apache' #35

**Comment by spdegabrielle** [comment]
Request to add to DrRacket: https://github.com/racket/racket/issues/2162
I commented https://github.com/racket/racket/issues/2162#issuecomment-618542800

> @spdegabrielle commented:
> @Syntacticlosure I think this is a great idea.
> I see two issues
> 
> I believe it should be in a license compatible with Racket - I believe Apache (and? Or?) MIT is correct
> DrRacket is an important teaching tool - it should be well tested prior to inclusion by teachers who use DrRacket - probably best asking for feedback on Racket-users mailing list.
> A possible starting point would be to make activating /installing it a configuration option ?
> 
> I think we need to start with the licence issue
> 
> Kind regards / 亲切的问候
> Stephen
>

**Comment by spdegabrielle** [comment]
Inviting new contributors @sorawee & @AlexKnauth 

This is a good plugin with lots a nice functionality. 

Do you see anything that could be improved to make it more suitable for inclusion in DrRacket?

**Comment by AlexKnauth** [comment]
One thing that would be nice is to have it follow DrRacket [Color Scheme][color-scheme] preferences based on [`framework:color-schemes`][framework:color-schemes]. The background for the file panel should follow the background preference [`framework:basic-canvas-background`][framework:basic-canvas-background] and the text should follow `framework:default-text-color` or something. When I change the color scheme of DrRacket the file panel should change with it.

I don't really know how this works, but maybe [`color-prefs:lookup-in-color-scheme`][color-prefs:lookup-in-color-scheme] would be helpful? Or maybe things like [`canvas:color<%>`][canvas:color<%>] could do at least part of this automatically? I don't know.

  [color-scheme]: https://docs.racket-lang.org/drracket/color-scheme.html
  [framework:color-schemes]: https://docs.racket-lang.org/framework/Color_Prefs.html#(idx._(gentag._2._(lib._scribblings%2Fframework%2Fframework..scrbl)))
  [framework:basic-canvas-background]: https://docs.racket-lang.org/framework/Canvas.html?#(idx._(gentag._1._(lib._scribblings%2Fframework%2Fframework..scrbl)))
  [color-prefs:lookup-in-color-scheme]: https://docs.racket-lang.org/framework/Color_Prefs.html#%28def._%28%28lib._framework%2Fmain..rkt%29._color-prefs~3alookup-in-color-scheme%29%29
  [canvas:color<%>]: https://docs.racket-lang.org/framework/Canvas.html#%28def._%28%28lib._framework%2Fmain..rkt%29._canvas~3acolor~3c~25~3e%29%29

**Comment by AlexKnauth** [comment]
So I've started trying to implement color scheme things in a branch called [`use-from-color-scheme`](https://github.com/AlexKnauth/files-viewer/tree/use-from-color-scheme), and so far I have the `dir-control%` part started, but I don't really know what the right thing to do would be for the `directory-list%` / `hierarchical-list%` part. However, one of the key functions which would probably be useful for this is:
```
;; use-from-color-scheme : Symbol [(U StyleDelta Color) -> Any] -> Void
(define (use-from-color-scheme sym fn)
  (fn (color-prefs:lookup-in-color-scheme sym))
  (color-prefs:register-color-scheme-entry-change-callback sym fn #t))
```
With example uses:
```
(use-from-color-scheme
 'framework:basic-canvas-background
 (λ (v) (set-canvas-background v)))
(use-from-color-scheme
 'framework:default-text-color
 (λ (v) (set! text-color v)))
(use-from-color-scheme
 'framework:paren-match-color
 (λ (v) (set! path-element-background v)))
```

**Comment by branneman** [comment]
I have moved @AlexKnauth's suggestion to [a new issue](https://github.com/MatrixForChange/files-viewer/issues/46) since thats a different topic if you ask me.

Getting back to the title of this issue, how do we get this to be part of the official DrRacket release? Having a file viewer/manager is a must-have for any modern IDE.

It seems there are issues in the official repos already: https://github.com/racket/drracket/issues/312 and https://github.com/racket/racket/issues/2162

**Comment by mfelleisen** [comment]
Is this available on the package catalog for people to try out or for DrR to include easily via some "extend with desirable packages" protocol?

**Comment by branneman** [comment]
> Is this available on the package catalog for people to try out or for DrR to include easily via some "extend with desirable packages" protocol?

@mfelleisen Yes, see [Installation instructions](https://github.com/MatrixForChange/files-viewer#installation) and [package catalog](https://pkgs.racket-lang.org/package/files-viewer): `raco pkg install files-viewer`

**Comment by mfelleisen** [comment]
Thanks. I will be working with it for a while. 

— How committed are you to maintenance? 
— How open are you to feature requests?

**Comment by Syntacticlosure** [comment]
I am so glad that this project has aroused many people's interests. However, currently, I need to pay more attentions on my studies and can't make persistent efforts on the project.

Anyone who wants to work on this project and has already been a member of racket org or PLT Team can reply me below and then become a collaborator of this project. Others can also make pull requests.

Both pull requests and suggestions are welcome.

P.S. My friend @yjqww6 's [drcomplete](https://github.com/yjqww6/drcomplete) is another amazing plugin for DrRacket.

**Comment by Syntacticlosure** [comment]
I am ready for the maintenance of this project now. 
This plugin seems to continue to work after several racket updates, so I don't know what features/changes should I add to it.
I want more suggestions/issues/PRs to work on.
@spdegabrielle @AlexKnauth @branneman @mfelleisen

**Comment by madkins23** [comment]
I was just looking at this project, thinking that I might cut a fork and try to add some minor functionality. I can file a few feature issues:
* no drag and drop to move files to another directory
* no way to create a new directory
* changing the current directory doesn't change the DrRacket current directory (so saving a new file became easier)
* updating workspace manager to track open tabs

These are things I thought that I might play with in a fork, but I'm happy to just write them up and let someone else do them.

**Comment by spdegabrielle** [comment]
> I am ready for the maintenance of this project now.
> This plugin seems to continue to work after several racket updates, so I don't know what features/changes should I add to it.
> I want more suggestions/issues/PRs to work on.
> @spdegabrielle @AlexKnauth @branneman @mfelleisen

Not really a feature or a problem but something I’ve wanted to try but don’t have the time for is a DrRacket ‘professional edition’ -  a DrRacket distribution configured with files-viewer and other plugins included and enabled, a selection of quickscripts,  dark mode and typed/racket by default.

Is that a silly idea?

https://docs.racket-lang.org/racket-build-guide/index.html

**Comment by mfelleisen** [comment]
@Syntacticlosure Sorry, this notification sailed right by me. -- I have used it ever since I responded and I am mostly happy with it. When it comes to such sw I always have immediate reactions ("wouldn't it be nice ..") but then I forget and use what I have. -- I am about to embark on a non-trivial project again so perhaps my wishes come back to me. -- In the meantime, thanks for working this out.

**Comment by madkins23** [comment]
I've posted a few minimal PRs which likely reflect my own idiosyncratic viewpoints:
* PR #50
* PR #51 
* PR #52 

I'm not going to fuss if any of them are ruled out. Overall the `files-viewer` plugin is a nice tool and I agree that it should be part of (or distributed with) `DrRacket`.

Most of my remaining issues seem to group together under the heading of "workspaces". I decided to [post](https://racket.discourse.group/t/drracket-workspaces/2025) all of my thoughts on the Racket discourse group for more general discussion.
